cirq/circuits/optimization_pass.py

```diff
@@ -145,6 +145,18 @@ def optimize_circuit(self, circuit: Circuit):
                     [e for e in range(i, i + opt.clear_span)])
                 new_operations = self.post_clean_up(
                     cast(Tuple[ops.Operation], opt.new_operations))
-                circuit.insert_at_frontier(new_operations, i, frontier)
 
+                flat_new_operations = tuple(ops.flatten_to_ops(new_operations))
+
+                new_qubits = set()
+                for flat_op in flat_new_operations:
+                    for q in flat_op.qubits:
+                        new_qubits.add(q)
```

Comment:
This logic to flatten the new ops and collecting the qubits from them is exactly repeated at the beginning of `insert_at_frontier`, I think we should just add this logic there: 

https://github.com/quantumlib/Cirq/blob/b3728aaf6d64477e0913015edec2223748aab54e/cirq/circuits/circuit.py#L1320-L1323

WDYT?

Reply:
You mean add the `new_qubits.issubset(set(opt.clear_qubits))` condition in `insert_at_frontier`? Though `insert_at_frontier` is currently only called from `optimization_pass`, I think the condition of checking that new qubits should be a subset of original qubits should logically belong in the `PointOptimizer` and not in `insert_at_frontier`.

Reply:
Yeah, that's what I meant. I am not confident that the abstractions are crisp here - the whole concept of a frontier only comes up in optimizers / compilers - so why does the Circuit class need to know about it. 
But you are right that this logic is an optimizer concern - also I totally missed that for my suggestion to work you'd have to pass `opt` as well to `insert_at_frontier` which makes the situation worse, leaking even more optimizer concepts to the Circuit class. 

So let's keep it this way, and we'll live with this small amount of duplication for now.